Fix #3934, #3958, and #3919: [RTL] High-fi Align TextViews, description text and toolbar marquee text.

veena14cs opened the pull request:
## Explanation

This PR fixes #3934, [#3958 ](https://github.com/oppia/oppia-android/issues/3958) and #3919 . Topic description and Continue description is right aligned. Marquee text for toolbar title is RTL compatible.

## Screenshot LTR and RTL

Mobile-

<img width="256" alt="Screenshot_1634368694" src="https://user-images.githubusercontent.com/16301028/137569389-cc29ca17-6556-4c77-a161-dcc0679d2bd8.png">......<img width="256" alt="Screenshot_1634368747" src="https://user-images.githubusercontent.com/16301028/137569377-5d0bffd3-c2a5-421a-89aa-ec23fd622e39.png">.....

<img width="144" alt="Screenshot_1634368653" src="https://user-images.githubusercontent.com/16301028/137569399-76836b86-8184-4f43-8afc-9ea11d9572ad.png">....<img width="144" alt="Screenshot_1634368861" src="https://user-images.githubusercontent.com/16301028/137569364-8bc97268-e079-4538-8ba2-78733beec6db.png">

<img width="144" alt="Screenshot_1634367255" src="https://user-images.githubusercontent.com/16301028/137569448-3cd6f730-2d27-45aa-b36a-bbafccc1dfba.png">...<img width="144" alt="Screenshot_1634362300" src="https://user-images.githubusercontent.com/16301028/137569462-828fd8e5-c098-46a6-8cd8-8a2486d1e160.png">

<img width="256" alt="Screenshot_1634368488" src="https://user-images.githubusercontent.com/16301028/137569425-a9859149-0c89-47c5-8522-b1501fab42aa.png">.....<img width="256" alt="Screenshot_1634368924" src="https://user-images.githubusercontent.com/16301028/137569351-623fbebb-9403-4e5f-94c9-60dd43b35ce6.png">

<img width="144" alt="Screenshot_1634861968" src="https://user-images.githubusercontent.com/16301028/138273049-955b4ebd-583b-4290-977a-460df45295b8.png">....<img width="144" alt="Screenshot_1634861995" src="https://user-images.githubusercontent.com/16301028/138273112-f21d3e7d-af46-4bff-9b79-a79270c1f408.png">

<img width="144" alt="Screenshot_1634861935" src="https://user-images.githubusercontent.com/16301028/138272881-bd22cb4c-b5a5-456d-bb16-78335ce5bf5e.png">.....<img width="144" alt="Screenshot_1634861681" src="https://user-images.githubusercontent.com/16301028/138272941-c18af236-96ae-4e77-a217-2142a89ae4ad.png">

<img width="144" alt="Screenshot_1634862870" src="https://user-images.githubusercontent.com/16301028/138274116-438af25c-0dd4-4d8e-b50d-b999b5bf361c.png">...<img width="144" alt="Screenshot_1634863171" src="https://user-images.githubusercontent.com/16301028/138274170-d1c144e8-719f-4f6a-995a-923156c7c096.png">

<img width="144" alt="Screenshot_1634863921" src="https://user-images.githubusercontent.com/16301028/138276395-4f6389a6-1452-4c93-9cec-2b1ba4a54037.png">....<img width="144" alt="Screenshot_1634863900" src="https://user-images.githubusercontent.com/16301028/138276466-2ae326c8-a7dd-4a22-bf7d-d146ec2738fa.png">
<img width="144" alt="Screenshot_1634884386" src="https://user-images.githubusercontent.com/16301028/138332321-00c750a3-7678-4b84-83cd-6dec0365b084.png">....<img width="144" alt="Screenshot_1634884365" src="https://user-images.githubusercontent.com/16301028/138332411-36eb9577-50c7-4575-b2aa-ac52a5946167.png">
<img width="144" alt="Screenshot_1634884390" src="https://user-images.githubusercontent.com/16301028/138332460-88328a29-7d46-4e01-8fab-0a1e11d8d26a.png">...<img width="144" alt="Screenshot_1634884362" src="https://user-images.githubusercontent.com/16301028/138332532-bb3ee146-6f8e-4424-9010-caffe09148fb.png">
<img width="144" alt="Screenshot_1634886847" src="https://user-images.githubusercontent.com/16301028/138339054-22614681-cd6b-4c6a-99f3-8f9ff1d6b627.png">......<img width="144" alt="Screenshot_1634886825" src="https://user-images.githubusercontent.com/16301028/138339129-6d164231-361c-459a-b26a-66d36ab8b945.png">

Tablet

<img width="256" alt="Screenshot_1634393196" src="https://user-images.githubusercontent.com/16301028/137569189-ec7297d1-487a-488e-b7cc-fd96a21d7617.png">....<img width="256" alt="Screenshot_1634393037" src="https://user-images.githubusercontent.com/16301028/137569326-d3e786a4-922c-4dcb-82fa-bb6f8f2c5e3d.png">

<img width="160" alt="Screenshot_1634393189" src="https://user-images.githubusercontent.com/16301028/137569214-f77ce1d1-1e4d-4867-8b52-f6c973c07eec.png">....<img width="160" alt="Screenshot_1634393045" src="https://user-images.githubusercontent.com/16301028/137569295-58f63065-60b0-4a78-b8b9-f4f962140f9d.png">

<img width="160" alt="Screenshot_1634393167" src="https://user-images.githubusercontent.com/16301028/137569226-a448c978-dcd3-46b5-b14e-be4176b61ccd.png">....<img width="160" alt="Screenshot_1634393103" src="https://user-images.githubusercontent.com/16301028/137569274-964a602c-80d1-46e7-aaf2-f0726097641f.png">

<img width="256" alt="Screenshot_1634393153" src="https://user-images.githubusercontent.com/16301028/137569248-acb6fe89-23d9-48b5-bff5-074332f9497d.png">.....<img width="256" alt="Screenshot_1634393113" src="https://user-images.githubusercontent.com/16301028/137569257-fdc04054-da2c-436f-b026-b387d0887c17.png">



## Marquee for Toolbar title LTR and RTL



https://user-images.githubusercontent.com/16301028/137569526-516a4ee5-1e33-46fb-8deb-ff95370369cb.mp4

https://user-images.githubusercontent.com/16301028/137569566-4e50d024-3d9a-4129-a2c4-4aeb8d89bab6.mp4


## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

BenHenning commented:
Sorry @veena14cs will need to review this after the weekend.

rt4914 reviewed (approved):
Using `textAlignment` is completely correct as per [this link](https://stackoverflow.com/questions/34275076/use-locale-ltr-rtl-for-gravity-in-textview) (Please read comments in this SO link for detailed information).

Thanks @veena14cs

veena14cs commented:
> Using `textAlignment` is completely correct as per [this link](https://stackoverflow.com/questions/34275076/use-locale-ltr-rtl-for-gravity-in-textview) (Please read comments in this SO link for detailed information).
> 
> Thanks @veena14cs

Thanks @rt4914

veena14cs committed:
Fix desrciption and marquee for RTL

veena14cs committed:
added test

veena14cs committed:
Update ResumeLessonFragmentTest.kt

veena14cs committed:
fixed tests

veena14cs committed:
fixed ellipsize

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentTest.kt:
```diff
@@ -197,6 +201,36 @@ class ResumeLessonFragmentTest {
   }
 
+  @Test
+  fun testResumeLessonFragment_lessonDescriptionIsInRTL_isDisplayedCorrectly() {
```
```suggestion
  fun testResumeLessonFragment_lessonDescriptionIsInRtl_isDisplayedCorrectly() {
```

Ditto elsewhere in PR (including for 'LTR' which should be 'Ltr').

veena14cs replied:
Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/info/TopicInfoFragmentTest.kt:
```diff
@@ -189,6 +198,45 @@ class TopicInfoFragmentTest {
+        R.id.topic_description_text_view
+      )
+      assertThat(topicDescriptionTextview.textAlignment)
+        .isEqualTo(View.TEXT_ALIGNMENT_VIEW_START)
```
Can this fit on one line? Prefer keeping it on one line if so. Ditto below & elsewhere in the PR.

veena14cs replied:
Done.

BenHenning commented on app/src/main/res/layout-land/resume_lesson_fragment.xml:
```diff
@@ -43,14 +43,14 @@
         android:ellipsize="end"
         android:fontFamily="sans-serif"
-        android:gravity="start"
+        android:textAlignment="viewStart"
```
I think it's really important that we ensure text alignment is always correct by default. I suggest adding a new Oppia theme for body TextViews that explicitly defines text alignment as view start, and set that as the default TextView style.

@rt4914 do you think that's reasonable? I'm worried at how easy it is to miss this, and that we might actually even yet have other TextViews that we didn't discover with the wrong text alignment. It'd be better to remove all these individual cases and centrally define it, by default.

veena14cs replied:
I created style for Marquee effect and description text.

BenHenning commented on app/src/main/res/layout-land/story_fragment.xml:
```diff
@@ -41,12 +42,12 @@
           android:layout_height="wrap_content"
           android:layout_marginEnd="20dp"
           android:ellipsize="marquee"
+          android:textAlignment="viewStart"
```
Similarly here. Can we solve this using a theme & styles, instead? There's another PR where a contributor is adding a new toolbar, and we should make sure this is set up such that we never have to manually fix the marquee directionality again.

/cc @rt4914

veena14cs replied:
Adde style with marquee effect for toolbar titles.

BenHenning reviewed (changes requested):
Thanks @veena14cs. This is a really nice solution. My main concern is that I want to make sure that we can fix this, by default, here & everywhere in the future for TextViews & Toolbars to make sure we never have to fix it again. I'm hoping that it's generally solvable via themes & styles. PTAL at my comments.

BenHenning commented:
Also, can #3920 be closed now that this PR seems like the better solution of the two?

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs commented:
> Also, can #3920 be closed now that this PR seems like the better solution of the two?

Yes, #3920 can be close as its getting resolved here.

veena14cs commented:
> Thanks @veena14cs. This is a really nice solution. My main concern is that I want to make sure that we can fix this, by default, here & everywhere in the future for TextViews & Toolbars to make sure we never have to fix it again. I'm hoping that it's generally solvable via themes & styles. PTAL at my comments.

I have updated the new style only to the files that are changed in this PR. PTAL. In future we need to update these style to other files too.

veena14cs committed:
Update ResumeLessonFragmentTest.kt

veena14cs committed:
fix lint

veena14cs committed:
add style for marquee and description text

veena14cs committed:
add width to the style

veena14cs committed:
Update ResumeLessonFragmentTest.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

BenHenning commented:
Sorry, will need to review this tomorrow.

veena14cs commented:
> Sorry, will need to review this tomorrow.

ACK

BenHenning commented on app/src/main/res/values/styles.xml:
```diff
@@ -239,9 +239,13 @@
   </style>
 
   <style name="Body" parent="TextView.Common1">
```
Does this cover *all* of the cases that we need it to? I think we should double check that other body-like texts are correctly included & use this style to ensure their alignment is correct.

veena14cs replied:
I will check all the textviews and update in the same PR.

BenHenning replied:
Can you follow up once you finish (to confirm that it's been done)?

veena14cs replied:
Its done. I have updated where it is required.

BenHenning replied:
Thanks!

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityTest.kt:
```diff
@@ -147,6 +152,46 @@ class RevisionCardActivityTest {
+          activity.findViewById(R.id.revision_card_toolbar_title)
+        ViewCompat.setLayoutDirection(revisionCardToolbarTitle, ViewCompat.LAYOUT_DIRECTION_LTR)
+
+        onView(withId(R.id.revision_card_toolbar_title)).perform(ViewActions.click())
```
Nit: please directly import click() for consistency with other tests.

veena14cs replied:
Done.

BenHenning reviewed (commented):
Thanks @veena14cs. I think this is looking really good. I mainly just had two comments, one of which is to ensure that we do a proper technical audit to ensure that we're covering all the cases that we want to for Body (both for cases where other texts should probably use Body, and for all existing cases using Body to ensure that the text direction change is correct for those cases). Please follow up once this has been done.

@rt4914 I think it might be good for you to take a quick follow-up pass to sign off on the theme/style changes since we want to make sure this doesn't cause any issues with ongoing night mode work.

BenHenning commented:
@veena14cs now that https://github.com/oppia/oppia-android/commit/43aee2e5e6e0b84fa7ea06f6288f185d738abcd5 is merged, can you please make sure that toolbar is also covered in this PR? I wasn't sure which was going to be merged first.

veena14cs commented:
> @veena14cs now that [43aee2e](https://github.com/oppia/oppia-android/commit/43aee2e5e6e0b84fa7ea06f6288f185d738abcd5) is merged, can you please make sure that toolbar is also covered in this PR? I wasn't sure which was going to be merged first.

The new changes will not affect [43aee2e](https://github.com/oppia/oppia-android/commit/43aee2e5e6e0b84fa7ea06f6288f185d738abcd5) as the change is for only the toolbar with changeable texts like topic,story or explorations that uses TextView to set title for the Toolbar.

veena14cs commented:
> Thanks @veena14cs. I think this is looking really good. I mainly just had two comments, one of which is to ensure that we do a proper technical audit to ensure that we're covering all the cases that we want to for Body (both for cases where other texts should probably use Body, and for all existing cases using Body to ensure that the text direction change is correct for those cases). Please follow up once this has been done.
> 
> @rt4914 I think it might be good for you to take a quick follow-up pass to sign off on the theme/style changes since we want to make sure this doesn't cause any issues with ongoing night mode work.

I have covered almost all the cases, wherever I found bug I have added screeshots in the PR description with the correction for LTR and RTL device. PTAL let me know if I have missed some thing.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

veena14cs committed:
fixed alignment where required

veena14cs committed:
fixed textalignment

veena14cs committed:
remove alignment from heading 2 style

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

BenHenning commented on app/src/main/res/layout/add_profile_activity.xml:
```diff
@@ -65,13 +65,15 @@
             android:textColor="@color/oppiaPrimaryText"
             android:textSize="12sp"
+            app:layout_constraintEnd_toEndOf="@+id/add_profile_activity_user_name"
+            app:layout_constraintStart_toStartOf="@+id/add_profile_activity_user_name"
```
Why this change? It seems unrelated to RTL.

veena14cs replied:
This is required to support LTR and RTL. otherwise its always left aligned.

BenHenning replied:
Ah, but doesn't this cause it to be centered? Is that what we want, or do we want it to be start-aligned?

veena14cs replied:
It has to be start-aligned and not at the center.

BenHenning replied:
Sorry to clarify, doesn't this constraint result in the text view being centered since it's start & end constrained on another view?

BenHenning reviewed (commented):
Thanks @veena14cs. This principally is looking excellent.

@rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?

BenHenning commented:
Also @veena14cs, please update to the latest develop & resolve the conflict when you get a chance (just to make sure this branch is staying up-to-date since it has far-reaching changes).

rt4914 commented on app/src/main/res/layout-land/revision_card_fragment.xml:
```diff
@@ -24,15 +24,11 @@
-        android:fontFamily="sans-serif"
-        android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        android:textColor="@color/oppiaPrimaryText" />
```
I think should should not be needed. It should already by part of the parent style. Please check and confirm.

veena14cs replied:
done.

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -239,9 +239,13 @@
     <item name="android:textSize">16sp</item>
     <item name="fontFamily">sans-serif</item>
+    <item name="android:textAlignment">viewStart</item>
+
```
Remove this blank and arrange the new attribute alphabetically.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -317,4 +321,23 @@
     <item name="android:textSize">14sp</item>
   </style>
+
+  <style name="MarqueeToolbarTitleTextView" parent="Widget.AppCompat.TextView">
```
Use parent as `TextView.Common1`

rt4914 replied:
Once this change is done then you can remove `layout_width` and `layout_height` from all views which use this style.

veena14cs replied:
Done. Is it necessary to remove or we can keep it.

rt4914 commented on app/src/main/res/layout/license_text_viewer_activity.xml:
```diff
@@ -20,28 +20,16 @@
-          android:singleLine="true"
-          android:textColor="@color/white"
-          android:textSize="20sp" />
+          android:layout_marginEnd="20dp"/>
```
Space missing
`android:layout_marginEnd="20dp" />`

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout/exploration_activity.xml:
```diff
@@ -40,23 +40,11 @@
-            android:singleLine="true"
-            android:textColor="@color/white"
-            android:textSize="20sp" />
+            android:layout_weight="1"/>
```
Space missing
`android:layout_weight="1" />`

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout/concept_card_fragment.xml:
```diff
@@ -55,21 +55,19 @@
-            android:fontFamily="sans-serif"
-            android:textColor="@color/oppiaPrimaryText"
-            android:textSize="16sp" />
+            android:textColor="@color/oppiaPrimaryText"/>
```
I thinks this is not needed.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp-port/topic_info_fragment.xml:
```diff
@@ -76,17 +76,16 @@
         android:paddingStart="120dp"
         android:paddingEnd="120dp"
         android:text="@{viewModel.topicDescription}"
         android:textColor="@color/oppiaPrimaryText"
```
This can also be removed.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp-port/topic_info_fragment.xml:
```diff
@@ -76,17 +76,16 @@
-        android:layout_width="wrap_content"
+        style="@style/Body"
+        android:layout_width="match_parent"
         android:layout_height="wrap_content"
```
This can be removed too.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp-land/topic_info_fragment.xml:
```diff
@@ -57,17 +57,14 @@
           android:paddingStart="64dp"
           android:paddingEnd="64dp"
           android:text="@{viewModel.topicDescription}"
           android:textColor="@color/oppiaPrimaryText"
```
This can be removed too.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout-land/topic_info_fragment.xml:
```diff
@@ -58,18 +58,14 @@
         android:paddingStart="72dp"
         android:paddingEnd="72dp"
         android:text="@{viewModel.topicDescription}"
         android:textColor="@color/oppiaPrimaryText"
```
This can be removed

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout-land/story_chapter_view.xml:
```diff
@@ -70,18 +70,14 @@
-            android:fontFamily="sans-serif"
-            android:textAlignment="viewStart"
             android:text="@{htmlContent}"
             android:textColor="@color/oppiaPrimaryText"
```
This can be removed

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
@veena14cs PTAL

rt4914 commented:
> Thanks @veena14cs. This principally is looking excellent.
> 
> @rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?

@BenHenning Unable to understand what exactly you mean here.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

rt4914 reviewed (approved):
LGTM, thanks.

veena14cs commented:
> LGTM, thanks.

Thanks @rt4914.

veena14cs committed:
fixed textstyle

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

BenHenning commented on app/src/main/res/layout/topic_info_fragment.xml:
```diff
@@ -87,7 +87,6 @@
         android:paddingStart="32dp"
         android:paddingEnd="32dp"
         android:text="@{viewModel.topicDescription}"
-        android:textColor="@color/oppiaPrimaryText"
```
It seems like the ``Body`` style isn't defining this as the textColor, should it be?

veena14cs replied:
Body style extends parent textview style i.e` TextView.Common1` here the textcolor is set already.

BenHenning replied:
Ah, thanks for clarifying.

BenHenning reviewed (commented):
Thanks @veena14cs! Just had a few comments, otherwise the fix looks excellent.

I think the main remaining point is the one brought up below regarding whether we should be ensuring all TextViews in the app have styles set up. I'll work with Rajat to make a decision on this quickly so that we can get the rest of this PR wrapped up.

> > Thanks @veena14cs. This principally is looking excellent.
> > @rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?
> 
> @BenHenning Unable to understand what exactly you mean here.

Sorry, some context was probably missing since @veena14cs & I have been chatted about this outside the PR. @rt4914 I'm wondering whether we should be ensuring that all TextViews have a predetermined style (like Body) so that we can centrally manage things like RTL for most scenarios just by ensuring that TextViews always have a style that has the correct LTR/RTL setup. We could even use a CI check in the future to enforce that TextViews properly use a predefined style (which is outside thes cope of this PR, but ensuring that all TextViews are styled seems within scope here). What are your thoughts?

rt4914 commented:
> Thanks @veena14cs! Just had a few comments, otherwise the fix looks excellent.
> 
> I think the main remaining point is the one brought up below regarding whether we should be ensuring all TextViews in the app have styles set up. I'll work with Rajat to make a decision on this quickly so that we can get the rest of this PR wrapped up.
> 
> > > Thanks @veena14cs. This principally is looking excellent.
> > > @rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?
> > 
> > 
> > @BenHenning Unable to understand what exactly you mean here.
> 
> Sorry, some context was probably missing since @veena14cs & I have been chatted about this outside the PR. @rt4914 I'm wondering whether we should be ensuring that all TextViews have a predetermined style (like Body) so that we can centrally manage things like RTL for most scenarios just by ensuring that TextViews always have a style that has the correct LTR/RTL setup. We could even use a CI check in the future to enforce that TextViews properly use a predefined style (which is outside thes cope of this PR, but ensuring that all TextViews are styled seems within scope here). What are your thoughts?

@BenHenning yes that makes sense and I completely agree with this as it would make things a lot more easier from RTL prospective.

veena14cs commented:
> > Thanks @veena14cs! Just had a few comments, otherwise the fix looks excellent.
> > I think the main remaining point is the one brought up below regarding whether we should be ensuring all TextViews in the app have styles set up. I'll work with Rajat to make a decision on this quickly so that we can get the rest of this PR wrapped up.
> > > > Thanks @veena14cs. This principally is looking excellent.
> > > > @rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?
> > > 
> > > 
> > > @BenHenning Unable to understand what exactly you mean here.
> > 
> > 
> > Sorry, some context was probably missing since @veena14cs & I have been chatted about this outside the PR. @rt4914 I'm wondering whether we should be ensuring that all TextViews have a predetermined style (like Body) so that we can centrally manage things like RTL for most scenarios just by ensuring that TextViews always have a style that has the correct LTR/RTL setup. We could even use a CI check in the future to enforce that TextViews properly use a predefined style (which is outside thes cope of this PR, but ensuring that all TextViews are styled seems within scope here). What are your thoughts?
> 
> @BenHenning yes that makes sense and I completely agree with this as it would make things a lot more easier from RTL prospective.

@BenHenning @rt4914 can we have a new issue and PR tracking this. As this PR is growing and if the correction on textviews looks good here lets make it merge ready and have another PR to ensure all textviews have style attribute.

rt4914 commented:
> > > Thanks @veena14cs! Just had a few comments, otherwise the fix looks excellent.
> > > I think the main remaining point is the one brought up below regarding whether we should be ensuring all TextViews in the app have styles set up. I'll work with Rajat to make a decision on this quickly so that we can get the rest of this PR wrapped up.
> > > > > Thanks @veena14cs. This principally is looking excellent.
> > > > > @rt4914 do you have any thoughts on to what extent we should be using styles for the non-body text views in this PR?
> > > > 
> > > > 
> > > > @BenHenning Unable to understand what exactly you mean here.
> > > 
> > > 
> > > Sorry, some context was probably missing since @veena14cs & I have been chatted about this outside the PR. @rt4914 I'm wondering whether we should be ensuring that all TextViews have a predetermined style (like Body) so that we can centrally manage things like RTL for most scenarios just by ensuring that TextViews always have a style that has the correct LTR/RTL setup. We could even use a CI check in the future to enforce that TextViews properly use a predefined style (which is outside thes cope of this PR, but ensuring that all TextViews are styled seems within scope here). What are your thoughts?
> > 
> > 
> > @BenHenning yes that makes sense and I completely agree with this as it would make things a lot more easier from RTL prospective.
> 
> @BenHenning @rt4914 can we have a new issue and PR tracking this. As this PR is growing and if the correction on textviews looks good here lets make it merge ready and have another PR to ensure all textviews have style attribute.

@veena14cs That does sound good to me. We can merge this PR and solve other things in separate PR. Make sure you file an issue for the same. Thanks.

veena14cs commented:
> > > @BenHenning @rt4914 can we have a new issue and PR tracking this. As this PR is growing and if the correction on textviews looks good here lets make it merge ready and have another PR to ensure all textviews have style attribute.
> > 
> > 
> > @veena14cs That does sound good to me. We can merge this PR and solve other things in separate PR. Make sure you file an issue for the same. Thanks.


 Thanks @rt4914. I have filed an issue #3971 tracking this. Once this PR is merged I can find it easy to solve this issue.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-textalignment

BenHenning reviewed (commented):
Thanks @veena14cs. Just have one comment to clarify, and then this PR will LGTM. Please ping me once you get a chance to review it--I'm happy to prioritize getting this PR merged today (my time).

> > > > @BenHenning @rt4914 can we have a new issue and PR tracking this. As this PR is growing and if the correction on textviews looks good here lets make it merge ready and have another PR to ensure all textviews have style attribute.
> > > 
> > > 
> > > @veena14cs That does sound good to me. We can merge this PR and solve other things in separate PR. Make sure you file an issue for the same. Thanks.
> 
> Thanks @rt4914. I have filed an issue #3971 tracking this. Once this PR is merged I can find it easy to solve this issue.

SGTM.

veena14cs commented:
> Thanks @veena14cs. Just have one comment to clarify, and then this PR will LGTM. Please ping me once you get a chance to review it--I'm happy to prioritize getting this PR merged today (my time).
> 

I have changed constraint to parent instead of on view. The width is set to `0dp` that is why it is not aligned in the center and it is always at left for LTR and right for RTL.

BenHenning commented:
> > Thanks @veena14cs. Just have one comment to clarify, and then this PR will LGTM. Please ping me once you get a chance to review it--I'm happy to prioritize getting this PR merged today (my time).
> 
> I have changed constraint to parent instead of on view. The width is set to `0dp` that is why it is not aligned in the center and it is always at left for LTR and right for RTL.

Thanks @veena14cs 

I'm having difficulty understanding this. Per https://stackoverflow.com/a/46692576 and https://stackoverflow.com/a/40262227 it's my understanding that setting start/end constraints to a view's parent (or another view) with 0dp width always centers that view relative to the other one (in the case of the parent, within the parent) since "0dp" forces the constraint's dimensions to be used as the width of the view.

That being said, because it's a TextView I think its text alignment is always start & top by default, so it seems that we don't need to do anything extra here. Can you confirm that we don't need to specify view alignment as start? And, if so, why is that not the case for other text views where we had to explicitly define it as viewStart?

veena14cs commented:
> > > Thanks @veena14cs. Just have one comment to clarify, and then this PR will LGTM. Please ping me once you get a chance to review it--I'm happy to prioritize getting this PR merged today (my time).
> > 
> > 
> > I have changed constraint to parent instead of on view. The width is set to `0dp` that is why it is not aligned in the center and it is always at left for LTR and right for RTL.
> 
> Thanks @veena14cs
> 
> I'm having difficulty understanding this. Per https://stackoverflow.com/a/46692576 and https://stackoverflow.com/a/40262227 it's my understanding that setting start/end constraints to a view's parent (or another view) with 0dp width always centers that view relative to the other one (in the case of the parent, within the parent) since "0dp" forces the constraint's dimensions to be used as the width of the view.
> 
> That being said, because it's a TextView I think its text alignment is always start & top by default, so it seems that we don't need to do anything extra here. Can you confirm that we don't need to specify view alignment as start? And, if so, why is that not the case for other text views where we had to explicitly define it as viewStart?

I think you got confused. Here I am not setting text to start/end but the entire TextView. You can check the screenshot below to fix the UI I have added constraint or I can add constraint bias with 0%.

**Old Screenshot LTR and RTL**
<img width="144" alt="Screenshot_1634863921" src="https://user-images.githubusercontent.com/16301028/138276395-4f6389a6-1452-4c93-9cec-2b1ba4a54037.png">....<img width="144" alt="Screenshot_1634864192" src="https://user-images.githubusercontent.com/16301028/138276849-ab6f46b3-4055-403b-8ac4-e601d2f2446e.png">

**New Screenshot LTR and RTL**

<img width="144" alt="Screenshot_1634863921" src="https://user-images.githubusercontent.com/16301028/138276395-4f6389a6-1452-4c93-9cec-2b1ba4a54037.png">....<img width="144" alt="Screenshot_1634863900" src="https://user-images.githubusercontent.com/16301028/138276466-2ae326c8-a7dd-4a22-bf7d-d146ec2738fa.png">


Earlier it wasn't constrained to parent or any other view and so it was left aligned even in RTL device. To fix this I added constraint.

BenHenning commented:
Got it. So the TextView is actually being centered now, and that's correct because the default alignment is viewStart. This would be easier to see if the bounds of the text view were rendered, or the constraints, but I think the screenshots help illustrate this point more clearly. Thanks @veena14cs

BenHenning reviewed (approved):
Per the discussion thread, this LGTM. Thanks @veena14cs!

veena14cs committed:
Update add_profile_activity.xml
